File: simpleblackjack.cpp
//
//  simpleblackjack.cpp
//  Simple Blackjack
//
//  Created by David Kopec on 8/29/16.
//  Ported from Chapter 6 of
//  Dart for Absolute Beginners by David Kopec
//  Copyright 2014
//

#include <string>
#include <vector>
#include <iostream>
#include <random>
#include <algorithm>

using namespace std;

// calculate the score of the cards Vector
int calculateScore(vector<string> &cards) {
    int score = 0;
    bool hasAce = false; // only one ace in Simple Blackjack
    
    // iterate through all cards
    for (string const& card: cards) {
        try {
            int cardValue = stoi(card);
            score += cardValue;
        } catch (std::invalid_argument& e) {
            // not a number card
            if (card == "A") { // it's an ace
                hasAce = true;
            } else { // it's a face card
                score += 10;
            }
        }
    }
    
    if (hasAce) {
        if ((score + 11) > 21) {  // don't let ace cause bust
            score += 1;
        } else {
            score += 11;
        }
    }
    
    return score;
}

// Print everyone's scores and decks
void printStatus(vector<string> &playerCards, vector<string> &dealerCards) {
    cout << "\n";  // blank line
    cout << "Player's total is " << calculateScore(playerCards) << ":\n";
    for (string const& card: playerCards) {
        cout << card << ", ";
    }
    cout << "\n";
    
    cout << "Dealer's total is " << calculateScore(dealerCards) << ":\n";
    for (string const& card: dealerCards) {
        cout << card << ", ";
    }
    cout << "\n";
}

// main game logic/loop
int main(int argc, const char * argv[]) {
    vector<string> deck = {"2", "3", "4", "5", "6", "7", "8", "9", "10", "J", "Q", "K", "A"};
    vector<string> playerCards = vector<string>();
    vector<string> dealerCards = vector<string>();
    
    // shuffle randomly the deck
    random_device rng;
    mt19937 urng(rng());
    shuffle(deck.begin(), deck.end(), urng);
    
    cout << "Dealer draws first card.\n";
    dealerCards.push_back(deck.back());
    deck.pop_back();
    cout << "Player receives two cards.\n";
    playerCards.push_back(deck.back());
    deck.pop_back();
    playerCards.push_back(deck.back());
    deck.pop_back();
    printStatus(playerCards, dealerCards);
    
    while (true) { // player decision loop
        cout << "Do you want to (H)it, (S)tay, or (Q)uit?\n";
        string selection;
        cin >> selection;
        // transform to upper case
        transform(selection.begin(), selection.end(),selection.begin(), ::toupper);
        
        if (selection == "H") { // hit
            playerCards.push_back(deck.back());
            deck.pop_back();
            printStatus(playerCards, dealerCards);
            
            if (calculateScore(playerCards) > 21) {
                cout << "You busted! You lose!\n";
                return 0;
            }
        } else if (selection == "S") { // stay
            break; // stop offering to hit, leave loop
        } else if (selection == "Q") { // quit
            return 0;
        }
    }
    
    cout << "Dealer draws rest of cards.\n";
    // keep drawing cards till 17
    while (calculateScore(dealerCards) < 17) {
        dealerCards.push_back(deck.back());
        deck.pop_back();
    }
    printStatus(playerCards, dealerCards);
    
    if (calculateScore(dealerCards) > 21) { // dealer bust
        cout << "Dealer busts! You win!\n";
    } else if (calculateScore(dealerCards) > calculateScore(playerCards)) {
        cout << "Dealer wins!\n";
    } else if (calculateScore(dealerCards) < calculateScore(playerCards)) {
        cout << "You win!\n";
    } else { // must be a tie by default
        cout << "It's a tie!\n";
    }
    return 0;
}

File: simple-blackjack.go

// Simple Blackjack for CSI 380
// This program play's a simple, single suit game of Blackjack
// against a computer dealer.
/*
Author: Brian Nguyen, Jacob Rose
Class: CSI 430-01
Assignment: Assignment 1
​Date Assigned:
Due Date: Jan 24, 2019
Description: Simple version of blackjack in go lang.
Certification of Authenticity: I certify that this is entirely my own work, except where I have given
fully‐documented references to the work of others. I understand the definition and
consequences of plagiarism and acknowledge that the assessor of this assignment
may, for the purpose of assessing this assignment:
‐ Reproduce this assignment and provide a copy to another member of academic
‐ staff; and/or Communicate a copy of this assignment to a plagiarism checking
‐ service (which may then retain a copy of this assignment on its database for the
‐ purpose of future plagiarism checking)
*/
package main

import (
	"fmt"
	"math/rand"
	"os"
	"strconv"
	//"strings"
	//"time"
)

//used in calculate 
var valueMap = map[string]int {
	"2" : 2,
	"3" : 3,
	"4" : 4,
	"5" : 5,
	"6" : 6,
	"7" : 7,
	"8" : 8,
	"9" : 9,
	"10":10,
	"J" :10,
	"Q" :10,
	"K" :10,
	"A" :11, //is checked in calculate score directly
}

// Move a card from deck to hand
func drawCard(hand *[]string, deck *[]string) {
	deckSize := len(*deck)-1
	tmpDeck := *deck
	topCard := tmpDeck[deckSize]
	tmpHand := append(*hand, topCard) //add last element from deck slice to the end of the hand
	*hand = tmpHand
	
	tmpDeck = tmpDeck[:len(*deck)-2] //should remove the last element from the deck
	*deck = tmpDeck
}

// Calculate the score of the hand
func calculateScore(hand []string) int {
	sum := 0
	aces := 0 //track aces and subtract them from the sum after if above 21
	for i := 0; i < len(hand); i++ {
		sum += valueMap[hand[i]] //get point value from valuemap constant defined above
		if(hand[i] == "A") {
			aces++ //count them and subtract later if needed
		}
	}
	for sum > 21 && aces > 0 { //if the score is above 21 and an ace was in the hand
		sum -= 10
		aces--
	}
	return sum
}

// Print everyone's scores and hands
func printStatus(playerCards, dealerCards []string) {
	fmt.Println("Player Cards:")
	for i:= 0; i < len(playerCards); i++ { //print all player cards
		fmt.Println(playerCards[i])
	}
	fmt.Println("Player Score: " + strconv.Itoa(calculateScore(playerCards)))
	fmt.Println("Dealer Cards:")
	for i:= 0; i < len(dealerCards); i++ { //print all dealer cards
		fmt.Println(dealerCards[i])
	}
	fmt.Println("Dealer Score: " + strconv.Itoa(calculateScore(dealerCards)))
}

// Entry point and main game loop
func main() {
	// YOU FILL IN HERE
	deck := []string{ "2" , "3" , "4" , "5" , "6" , "7" , "8" , "9" , "10" , "J", "Q" , "K" ,"A" }
	var playerCards []string
	var dealerCards []string

	//Shuffle randomly the deck
	//rand.seed(time.Now().UnixNano())
	//Fisher-Yates Shuffle
	for i := len(deck) - 1; i > 0; i-- {
	j := rand.Intn(i + 1)
		deck[i], deck[j] = deck[j], deck[i]
	}
	  fmt.Printf("Dealer draws first card.")
		  drawCard(&dealerCards, &deck)
		  fmt.Printf("Player receives two cards.")
		  drawCard(&playerCards, &deck)
		  drawCard(&playerCards, &deck)
		  printStatus(playerCards, dealerCards)

		  //player decision loop
		  var decision bool = true
		  for decision == true {
			  fmt.Printf("Do you want to (H)it, (S)tay, or (Q)uit?")
				  var response string
				  fmt.Scanln(&response)
				  fmt.Printf(response)
				  if response == "H" || response == "h" {
					  drawCard(&playerCards, &deck)
						  printStatus(playerCards, dealerCards)
						  if calculateScore(playerCards) > 21{
							fmt.Printf("You busted! You lose!")
							os.Exit(0)
						  }
				  	} else if response == "S" || response == "s"{
					  fmt.Printf("You chose to stay")
						  printStatus(playerCards, dealerCards)
						  decision = false
				  	} else {
					  os.Exit(0)
				  	}

		  }

	  //Dealer's turn
	  fmt.Printf("Dealer draws rest of cards.")
		  for calculateScore(dealerCards) < 17	{
			  drawCard(&dealerCards, &deck)
		  }
	  printStatus(playerCards, dealerCards)

		  if calculateScore(dealerCards) > 21 {
			  fmt.Printf("Dealer busts! You win!")
		  } else if calculateScore(dealerCards) < calculateScore(playerCards) {
			  fmt.Printf("Dealer wins!")
		  } else {
			  fmt.Printf("It's a tie!")
		  }
	os.Exit(0)
}
